### PR TITLE
BTRX-408 DUOS: DAR Confirmation Email to Researcher 

### PR DESCRIPTION
**Pull request by vvicario**
Includes UID, name, and data use limitations of the datasets in email with attestation reminder

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [X] I've followed [the instructions](https://github.com/DataBiosphere/consent/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/DataBiosphere/consent/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing
- [ ] I've updated Swagger to reflect any API changes

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/java/org/broadinstitute/consent/http/db/ElectionDAO.java [hunk 1]
```diff
@@ -281,4 +281,6 @@ void updateElectionById(@Bind("electionId") Integer electionId,
     @Mapper(DateMapper.class)
     Date findApprovalAccessElectionDate(@Bind("referenceId") String referenceId);
 
-}
\ No newline at end of file
+    @SqlQuery("select  v.vote from vote v where v.electionId = :electionId and v.type = 'FINAL'")
+    Boolean findFinalAccessVote(@Bind("electionId") Integer electionId);
+}
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/db/ElectionDAO.java:285]: Not really part of this PR, but `import org.skife.jdbi.v2.sqlobject.customizers.RegisterColumnMapper;` is unused

File: src/main/java/org/broadinstitute/consent/http/mail/freemarker/FreeMarkerTemplateHelper.java [hunk 2]
```diff
@@ -85,6 +86,18 @@ public Writer getHelpReportTemplate(HelpReport helpReport, String serverUrl) thr
         return generateHelpReportTemplate(helpReport, serverUrl + HELP_REPORT_URL, temp);
     }
 
+    public Writer getResearcherDarApprovedTemplate(String darCode, String researcherName, List<DatasetMailDTO> datasets, String dataUseRestriction, String email) throws IOException, TemplateException {
+        Template temp = freeMarkerConfig.getTemplate("researcher-dar-approved.html");
+        return generateResearcherApprovedTemplate(datasets, dataUseRestriction, darCode, researcherName, email, temp);
+    }
+    private Writer generateResearcherApprovedTemplate(List<DatasetMailDTO> datasets,  String dataUseRestriction, String darCode, String researcherName, String email, Template temp) throws IOException, TemplateException {
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/mail/freemarker/FreeMarkerTemplateHelper.java:94]: nit: add a newline between these methods

File: src/main/java/org/broadinstitute/consent/http/mail/freemarker/ResearcherDarApprovedModel.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    private String dataUseRestriction;
+    private String researcherEmail;
+
+    public ResearcherDarApprovedModel(String researcherName, String darCode, List<DatasetMailDTO> datasets, String dataUseRestriction, String researcherEmail) {
+        this.researcherName = researcherName;
+        this.darCode = darCode;
+        this.datasets = datasets;
+        this.dataUseRestriction = dataUseRestriction;
+        this.researcherEmail = researcherEmail;
+    }
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/mail/freemarker/ResearcherDarApprovedModel.java]: I'm definitely not asking that you change this, but I want to suggest what I think is a more readable pattern. Instead of using large constructors like this, I really like [the builder pattern](https://www.geeksforgeeks.org/builder-pattern-in-java/) where we would instead make a new `ResearcherDarApprovedModel` like this:
```
new ResearcherDarApprovedModel()
    .researcherName(name)
    .darCode(code)
    .dataSets(dataSets)
    .dataUseRestriction(useRestriction)
    .researcherEmail(email)
```
I find this pattern is much nicer to use, especially in tests.

Food for thought.

File: src/main/java/org/broadinstitute/consent/http/mail/message/ResearcherApprovedMessage.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+public class ResearcherApprovedMessage extends MailMessage {
+    private final String APPROVED_DAR = "Your DUOS Data Access Request Results";
+
+    public List<Mail> researcherApprovedMessage(Set<String> toAddresses, String fromAddress, Writer template, String darCode) throws MessagingException {
+        return generateEmailMessages(toAddresses, fromAddress, template, darCode, null);
+    }
+
+    @Override
+    String assignSubject(String referenceId, String type) {
+        return String.format(APPROVED_DAR, referenceId);
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/mail/message/ResearcherApprovedMessage.java]: `APPROVED_DAR` doesn't have any replacement arguments so `referenceId` will get dropped.

File: src/main/java/org/broadinstitute/consent/http/service/DatabaseElectionAPI.java [hunk 5]
```diff
@@ -701,5 +706,18 @@ private void updateFinalVote(Election rec, Integer electionId) {
         }
     }
 
-
+    private void sendResearcherNotification(String referenceId) throws Exception {
+        BasicDBObject query = new BasicDBObject(DarConstants.ID, new ObjectId(referenceId));
+        Document dar = mongo.getDataAccessRequestCollection().find(query).first();
+        List<Integer> dataSetIdList = dar.get(DarConstants.DATASET_ID, List.class);
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/DatabaseElectionAPI.java]: I've seen this pattern a few times using an unsafe cast of `List` to `List<T>`. We should try to wrap this assignment into a cast-safe utility method somewhere and add some TODOs about using it.

File: src/main/resources/freemarker/researcher-dar-approved.html [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                <p>Thank you,<br>the DUOS team</p>
+            </td>
+        </tr>
+    </table>
+    <table width ="600" style="font-family: 'Roboto', sans-serif ; margin-top: 15px;">
+        <tr align="center" style="font-family: 'Roboto', sans-serif ; color: #999999;">
+            <td align="center" valign="middle" style="padding: 0 10px 0 10px; border-collapse: collapse; font-family: 'Roboto', sans-serif; font-size: 14px; color: #999999; text-align: center; line-height: 21px; display: inline-block; vertical-align: middle;">	&#169; Broad Institute </td>
+        </tr>
+    </table>
+</center>
```
Comment (rushtong) [inline src/main/resources/freemarker/researcher-dar-approved.html:75]: `center` is a deprecated tag - not sure how this works in email. For PRs like this, it's also helpful to have screenshots of the emails that go out. It makes it a lot easier to review the PR.

File: src/test/java/org/broadinstitute/consent/http/mail/message/ResearcherApprovedMessageTest.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void testMessageSubject() throws MessagingException {
+        List<Mail> messages = new ResearcherApprovedMessage().researcherApprovedMessage(Collections.singleton("to@address.com"), "from@address.com", template, "DAR-123");
+        assertTrue(messages.get(0).getSubject().equals("Your DUOS Data Access Request Results"));
```
Comment (rushtong) [inline src/test/java/org/broadinstitute/consent/http/mail/message/ResearcherApprovedMessageTest.java]: Can be simplified:
```
assertEquals("Your DUOS Data Access Request Results", messages.get(0).getSubject());
```

File: src/main/java/org/broadinstitute/consent/http/mail/MailService.java [hunk 8]
```diff
@@ -136,4 +137,10 @@ public void sendNewHelpReportMessage(Set<String> usersAddress,  Writer template,
         sendMessages(messages);
     }
 
-}
\ No newline at end of file
+    @Override
+    public void sendNewResearcherApprovedMessage(Set<String> researcherEmails, Writer template, String darCode) throws MessagingException {
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/mail/MailService.java:141]: Could use some unit testing to ensure that it runs successfully and one for failing with a `MessagingException`

**Review by rushtong — CHANGES_REQUESTED** [review]
See comments inline. Could use some more unit tests, especially for the API implementation classes.

File: src/main/resources/freemarker/researcher-dar-approved.html (hunk 6 above)
Comment (nlopezzalba) [reply src/main/resources/freemarker/researcher-dar-approved.html:75]: Hi @rushtong, these are the screenshots of each email template.
researcher-dar-approved
dataowner-dar-approved
I'm aware align="center" is deprecated, unfortunately some old browsers and email services like Outlook still need it :(

Comment (rushtong) [reply src/main/resources/freemarker/researcher-dar-approved.html:75]: Thank you @nlopezzalba ! Disregard my `center` comment then - makes sense that you need it.

Comment (nlopezzalba) [reply src/main/resources/freemarker/researcher-dar-approved.html:75]: sure, thank you!

**Comment by vvicario** [comment]
@rushtong  I worked on your feedback! Please let me know if you have any other comment

File: src/main/java/org/broadinstitute/consent/http/util/DarUtil.java [hunk 9]
```diff
@@ -35,4 +36,15 @@ public static boolean requiresManualReview(Document dar) throws IOException {
...
+
+    public static  List<Integer> getIntegerList(Document dar, String key) {
+        List<Integer> result = new ArrayList<>();
+        List datasets = dar.get(key, List.class);
+        for(Object value : datasets){
+            if(value instanceof Integer) {
+                result.add((Integer)value);
+            }
+        }
+        return result;
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/util/DarUtil.java]: I know that you know I will always favor a lambda for this :-) This could be a lot simpler:
```
        List<Object> datasets = dar.get(key, List.class);
        return datasets.
                stream().
                filter(Integer.class::isInstance).
                map(Integer.class::cast).
                collect(Collectors.toList());
```

**Review by rushtong — APPROVED** [review]
One suggestion, but otherwise looks good! 👍
